cores/arduino/UARTClass.cpp

```diff
@@ -183,8 +183,7 @@ size_t UARTClass::write( const uint8_t uc_data )
   {
     // If busy we buffer
     int l = (_tx_buffer->_iHead + 1) % UART_BUFFER_SIZE;
-    while (_tx_buffer->_iTail == l)
-      ; // Spin locks if we're about to overwrite the buffer. This continues once the data is sent
+    while (*(volatile int*)&(_tx_buffer->_iTail) == l); // Spin locks if we're about to overwrite the buffer. This continues once the data is sent
```

Comment:
The fact of the matter is that any variable that can be changed by one task (eg. foreground) and being polled by another (eg. background) needs the actual access to the variable itself, no staging of the variable allowed.  Thus, no compiler optimization can be used when accessing this kind of variable and that means it can only be volatile variable.  This is to guarantee that any code will have access to the real value of this kind of variable without being aware of the fact that the variable can be changed in real time.

I concur that the head, tail, and overflow variable in RingBuffer class to be made volatile.